pkg/admissionpolicy/mutate.go

```diff
@@ -112,23 +126,310 @@ func mutateResource(
+			ruleResp := engineapi.RuleSkip(policy.GetName(), engineapi.Mutation, "mutation had no effect", nil)
+			policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
+			return engineResponse.WithPolicyResponse(policyResp), nil
 		}
-		versionedAttributes.Dirty = true
+
 		versionedAttributes.VersionedObject = newVersionedObject
+
 	}
+
 	patchedResource, err := celutils.ConvertObjectToUnstructured(versionedAttributes.VersionedObject)
 	if err != nil {
-		return engineResponse, err
+		ruleResp := engineapi.RuleError(policy.GetName(), engineapi.Mutation, err.Error(), nil, nil)
+		policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
+		return engineResponse.WithPolicyResponse(policyResp), nil
 	}
 
 	ruleResp := engineapi.RulePass(policy.GetName(), engineapi.Mutation, "", nil)
+	if binding != nil {
+		ruleResp = ruleResp.WithMutatingBinding(binding)
+	}
 	policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
 	engineResponse = engineResponse.
 		WithPatchedResource(*patchedResource).
 		WithPolicyResponse(policyResp)
 
 	return engineResponse, nil
 }
+
+// MutatingPolicyData holds a MAP and its associated MAPBs
+type MutatingPolicyData struct {
+	definition admissionregistrationv1alpha1.MutatingAdmissionPolicy
+	bindings   []admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding
+}
+
+// NewMutatingPolicyData initializes a MAP wrapper with no bindings
+func NewMutatingPolicyData(p admissionregistrationv1alpha1.MutatingAdmissionPolicy) MutatingPolicyData {
+	return MutatingPolicyData{definition: p}
+}
+
+// AddBinding appends a MAPB to the policy data
+func (m *MutatingPolicyData) AddBinding(b admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding) {
+	m.bindings = append(m.bindings, b)
+}
+
+func toGVR(gvk schema.GroupVersionKind) schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    gvk.Group,
+		Version:  gvk.Version,
+		Resource: strings.ToLower(gvk.Kind) + "s",
+	}
+}
+
+func Mutate(
+	data MutatingPolicyData,
+	resource unstructured.Unstructured,
+	client dclient.Interface,
+	namespaceSelectorMap map[string]map[string]string,
+) (engineapi.EngineResponse, error) {
+	var emptyResp engineapi.EngineResponse
+
+	resPath := fmt.Sprintf("%s/%s/%s", resource.GetNamespace(), resource.GetKind(), resource.GetName())
+	policy := data.definition
+	bindings := data.bindings
+
+	gvk := resource.GroupVersionKind()
+	gvr := schema.GroupVersionResource{
+		Group:    gvk.Group,
+		Version:  gvk.Version,
+		Resource: strings.ToLower(gvk.Kind) + "s",
+	}
+
+	a := admission.NewAttributesRecord(
+		resource.DeepCopyObject(), nil,
+		gvk,
+		resource.GetNamespace(), resource.GetName(),
+		gvr, "", admission.Create, nil, false, nil,
+	)
+
+	// no bindings
+	if len(bindings) == 0 {
+		//admissionregistrationv1.MatchResources(*policy.Spec.MatchConstraints)
+		isMatch, err := matches(a, namespaceSelectorMap, convertMatchResources(*policy.Spec.MatchConstraints))
```

Comment:
This function is removed as it was a duplicate of another one. You can use this one:
https://github.com/kyverno/kyverno/blob/b176532d6a10ee051dfb87f3771d8424ba47b9e9/pkg/cel/matching/matcher.go#L27
Here is how I used it in VAPs:
https://github.com/kyverno/kyverno/blob/b176532d6a10ee051dfb87f3771d8424ba47b9e9/pkg/admissionpolicy/validate.go#L105-L112